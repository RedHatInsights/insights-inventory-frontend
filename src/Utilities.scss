// Sass Imports

@import 'Utilities/_variables.scss';
@import 'Utilities/_colors.scss';
@import 'Utilities/_helpers.scss';
@import 'Utilities/_mixins.scss';

@import '~@redhat-cloud-services/frontend-components/index.css';
