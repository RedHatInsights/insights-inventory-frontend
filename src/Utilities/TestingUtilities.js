import React from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';

export const mountWithRouter = (Component, initialEntries) => {
  const wrapper = mount(
    <MemoryRouter initialEntriest={initialEntries}>{Component}</MemoryRouter>
  );

  return wrapper;
};

export const TestWrapper = ({
  children,
  routerProps = { initialEntries: ['/'] },
  path,
  store = {},
}) => {
  return (
    <MemoryRouter {...routerProps}>
      <Provider store={store}>
        {path ? (
          <Routes>
            <Route path={path} element={children} />
          </Routes>
        ) : (
          children
        )}
      </Provider>
    </MemoryRouter>
  );
};

TestWrapper.propTypes = {
  children: PropTypes.any.isRequired,
  routerProps: PropTypes.shape({
    initialEntries: PropTypes.arrayOf(PropTypes.string),
  }),
  path: PropTypes.string,
  store: PropTypes.any,
};
