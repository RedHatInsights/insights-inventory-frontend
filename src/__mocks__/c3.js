// eslint-disable-next-line no-undef
module.exports = () => 'c3';

