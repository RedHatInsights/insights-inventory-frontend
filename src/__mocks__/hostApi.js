import { hosts, tags } from '../api/api';
import MockAdapter from 'axios-mock-adapter';

export const mock = new MockAdapter(hosts.axios, { onNoMatch: 'throwException' });
export const mockTags = new MockAdapter(tags.axios, { onNoMatch: 'throwException' });
