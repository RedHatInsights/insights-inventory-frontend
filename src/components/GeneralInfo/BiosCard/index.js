export { default as BiosCard } from './BiosCard';
export { default } from './BiosCard';
