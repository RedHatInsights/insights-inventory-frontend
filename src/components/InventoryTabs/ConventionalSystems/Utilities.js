import flatMap from 'lodash/flatMap';
import {
  HOST_GROUP_CHIP,
  RHCD_FILTER_KEY,
  UPDATE_METHOD_KEY,
} from '../../../Utilities/constants';

const mapTags = ({ category, values }) =>
  values.map(
    ({ tagKey, value }) =>
      `${category ? `${category}/` : ''}${tagKey}${value ? `=${value}` : ''}`
  );

const filterMapper = {
  staleFilter: ({ staleFilter }, searchParams) =>
    staleFilter.forEach((item) => searchParams.append('status', item)),
  osFilter: ({ osFilter }, searchParams) => {
    // TODO This is very hackish. There should be a schema/feature on how the values translate into params
    const osParams = Object.entries(
      Object.values(osFilter).reduce((acc, item) => ({ ...acc, ...item }), {})
    )
      .filter(([, value]) => value === true)
      .map(([key]) => key);

    osParams.forEach((item) => {
      const keyParts = item.split('-');
      const [osName, value] = [
        keyParts.slice(0, keyParts.length - 2).join(' '),
        keyParts[keyParts.length - 1],
      ];

      if (value && osName) {
        searchParams.append('operating_system', `${osName}${value}`);
      }
    });
  },
  registeredWithFilter: ({ registeredWithFilter }, searchParams) =>
    registeredWithFilter?.forEach((item) =>
      searchParams.append('source', item)
    ),
  value: ({ value, filter }, searchParams) =>
    value === 'hostname_or_id' &&
    Boolean(filter) &&
    searchParams.append('hostname_or_id', filter),
  tagFilters: ({ tagFilters }, searchParams) =>
    tagFilters?.length > 0 &&
    searchParams.append('tags', flatMap(tagFilters, mapTags)),
  rhcdFilter: ({ rhcdFilter }, searchParams) =>
    rhcdFilter?.forEach((item) => searchParams.append(RHCD_FILTER_KEY, item)),
  lastSeenFilter: ({ lastSeenFilter }, searchParams) =>
    Object.keys(lastSeenFilter || {})?.forEach(
      (item) =>
        item === 'mark' &&
        searchParams.append('last_seen', lastSeenFilter[item])
    ),
  updateMethodFilter: ({ updateMethodFilter }, searchParams) =>
    updateMethodFilter?.forEach((item) =>
      searchParams.append(UPDATE_METHOD_KEY, item)
    ),
  hostGroupFilter: ({ hostGroupFilter }, searchParams) =>
    hostGroupFilter?.forEach((item) =>
      searchParams.append(HOST_GROUP_CHIP, item)
    ),
};

// TODO It might be better to not use a SearchParam instance and mutate it
// This should rather just return a plain object that can be
// applied with react routers useSearch params.
export const calculateFilters = (searchParams, filters = []) => {
  filters.forEach((filter) => {
    Object.keys(filter).forEach((key) => {
      filterMapper?.[key]?.(filter, searchParams);
    });
  });
  return searchParams;
};

export const calculatePagination = (searchParams, page, perPage) => {
  const currSearch = new URLSearchParams(location.search);
  const newPage = page !== undefined ? page : currSearch.get('page');
  const newPerPage =
    perPage !== undefined ? perPage : currSearch.get('per_page');
  !isNaN(parseInt(newPage)) && searchParams.append('page', newPage);
  !isNaN(parseInt(newPerPage)) && searchParams.append('per_page', newPerPage);
};
