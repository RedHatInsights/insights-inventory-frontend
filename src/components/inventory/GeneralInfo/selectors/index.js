export { default } from './selectors';
export * from './selectors';
