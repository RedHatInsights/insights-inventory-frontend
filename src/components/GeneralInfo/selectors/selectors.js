/* eslint-disable camelcase */
import { registered } from '../../../Utilities/index';
import { verifyCollectorStaleness } from '../../../Utilities/sharedFunctions';
function safeParser(toParse, key) {
  try {
    return JSON.parse(toParse);
  } catch {
    return { [key]: toParse };
  }
}

export const propertiesSelector = (
  {
    number_of_cpus,
    number_of_sockets,
    cores_per_socket,
    ramSize,
    disk_devices,
    sap_sids,
    system_purpose,
    cpu_flags,
  } = {},
  { facts } = {}
) => ({
  cpuNumber: number_of_cpus || facts?.rhsm?.CPU_CORES,
  sockets: number_of_sockets || facts?.rhsm?.CPU_SOCKETS,
  coresPerSocket:
    cores_per_socket ||
    (facts?.rhsm?.CPU_CORES &&
      facts?.rhsm?.CPU_CORES &&
      Number(facts?.rhsm?.CPU_CORES, 10) /
        Number(facts?.rhsm?.CPU_SOCKETS, 10)),
  ramSize: ramSize || (facts?.rhsm?.MEMORY && `${facts?.rhsm?.MEMORY} GB`),
  storage:
    disk_devices &&
    disk_devices.map(({ device, label, mount_point, options, type }) => ({
      ...(device && safeParser(device, 'device')),
      label,
      ...(mount_point && safeParser(mount_point, 'mountpoint')),
      ...(options && safeParser(options, 'options')),
      ...(type && safeParser(type, 'mounttype')),
    })),
  sapIds: sap_sids,
  systemPurpose: system_purpose?.usage,
  cpuFlags: cpu_flags,
});

export const operatingSystem = (
  {
    arch,
    operating_system,
    os_kernel_version,
    last_boot_time,
    kernel_modules,
    system_update_method,
  } = {},
  { facts } = {}
) => ({
  release: operating_system,
  kernelRelease: os_kernel_version,
  architecture: arch || facts?.rhsm?.ARCHITECTURE,
  bootTime: last_boot_time,
  kernelModules: kernel_modules,
  systemUpdateMethod: system_update_method,
});

export const biosSelector = ({
  bios_vendor,
  bios_version,
  bios_release_date,
} = {}) => ({
  vendor: bios_vendor,
  version: bios_version,
  releaseDate: bios_release_date,
});

export const bootcSelector = ({ bootc_status } = {}) => ({
  bootedImage: bootc_status?.booted?.image,
  bootedImageDigest: bootc_status?.booted?.image_digest,
  stagedImage: bootc_status?.staged?.image,
  stagedImageDigest: bootc_status?.staged?.image_digest,
  rollbackImage: bootc_status?.rollback?.image,
  rollbackImageDigest: bootc_status?.rollback?.image_digest,
});

export const infrastructureSelector = (
  {
    infrastructure_type,
    infrastructure_vendor,
    public_ipv4_addresses,
    public_dns,
    network = {},
  } = {},
  { facts } = {}
) => ({
  type:
    infrastructure_type ||
    (facts?.rhsm?.IS_VIRTUAL !== undefined &&
      (facts?.rhsm?.IS_VIRTUAL ? 'virtual' : 'physical')) ||
    undefined,
  vendor: infrastructure_vendor,
  public_ipv4_addresses: public_ipv4_addresses,
  ipv4: network.ipv4,
  ipv6: network.ipv6,
  nics: network.interfaces,
  fqdn: public_dns,
});

export const configurationSelector = ({
  installed_packages,
  enabled_services,
  running_processes,
  repositories,
} = {}) => ({
  packages: installed_packages,
  services: enabled_services,
  processes: running_processes,
  repositories,
});

export const collectionInformationSelector = ({
  insights_client_version,
  insights_egg_version,
} = {}) => ({
  client: insights_client_version,
  egg: insights_egg_version,
});

export const getCollectorStatus = (collectorStaleness) => {
  return collectorStaleness
    ? verifyCollectorStaleness(collectorStaleness) !== 'Fresh'
      ? 'Stale'
      : 'Active'
    : 'N/A';
};

export const getDefaultCollectors = (entity) =>
  registered
    ?.filter((reporter) => reporter.label !== 'insights-client not connected')
    .map((reporter) => ({
      name: reporter.label,
      status: getCollectorStatus(
        entity?.per_reporter_staleness[reporter.value]
      ),
      updated: entity?.per_reporter_staleness[reporter.value]?.last_check_in,
      details: {
        name: reporter.idName,
        id: entity?.[reporter.idValue],
      },
    }));

export const systemStatus = ({ stale_timestamp } = {}) => ({
  stale: new Date() > new Date(stale_timestamp),
});
