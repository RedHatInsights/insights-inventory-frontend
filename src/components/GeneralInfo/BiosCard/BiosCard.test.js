import { render } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';
import { TestWrapper } from '../../../Utilities/TestingUtilities';
import { biosTest } from '../../../__mocks__/selectors';
import BiosCard from './BiosCard';

describe('BiosCard', () => {
  let initialState;
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore();
    initialState = {
      systemProfileStore: {
        systemProfile: {
          loaded: true,
          ...biosTest,
          cpu_flags: ['one'],
        },
      },
    };
  });

  it('should render correctly - no data', () => {
    const store = mockStore({ systemProfileStore: {} });
    const view = render(
      <TestWrapper store={store}>
        <BiosCard />
      </TestWrapper>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    const store = mockStore(initialState);
    const view = render(
      <TestWrapper store={store}>
        <BiosCard />
      </TestWrapper>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render correctly with data - wrong date', () => {
    const store = mockStore({
      systemProfileStore: {
        systemProfile: {
          loaded: true,
          ...biosTest,
          bios_release_date: 'test',
          cpu_flags: ['one'],
        },
      },
    });
    const view = render(
      <TestWrapper store={store}>
        <BiosCard />
      </TestWrapper>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });

  ['hasVendor', 'hasVersion', 'hasReleaseDate'].map((item) =>
    it(`should not render ${item}`, () => {
      const store = mockStore(initialState);
      const view = render(
        <TestWrapper store={store}>
          <BiosCard {...{ [item]: false }} />
        </TestWrapper>
      );
      expect(view.asFragment()).toMatchSnapshot();
    })
  );

  it('should render extra', () => {
    const store = mockStore(initialState);
    const view = render(
      <TestWrapper store={store}>
        <BiosCard
          extra={[
            { title: 'something', value: 'test' },
            {
              title: 'with click',
              value: '1 tests',
              onClick: (_e, handleClick) =>
                handleClick('Something', {}, 'small'),
            },
          ]}
        />
      </TestWrapper>
    );

    expect(view.asFragment()).toMatchSnapshot();
  });
});
