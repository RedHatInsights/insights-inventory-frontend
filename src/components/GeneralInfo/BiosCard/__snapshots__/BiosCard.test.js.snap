// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BiosCard should not render hasReleaseDate 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-6"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-6"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              test-vendor
            </dd>
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              test-version
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should not render hasVendor 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-4"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-4"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              test-version
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              01 Apr 2014
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should not render hasVersion 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-5"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-5"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              test-vendor
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              01 Apr 2014
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should render correctly - no data 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-1"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-1"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              <div
                class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
              >
                <span
                  class="pf-u-screen-reader"
                />
              </div>
            </dd>
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              <div
                class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
              >
                <span
                  class="pf-u-screen-reader"
                />
              </div>
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              <div
                class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__sm"
              >
                <span
                  class="pf-u-screen-reader"
                />
              </div>
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should render correctly with data - wrong date 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-3"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-3"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              test-vendor
            </dd>
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              test-version
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              Not available
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should render correctly with data 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-2"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-2"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              test-vendor
            </dd>
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              test-version
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              01 Apr 2014
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;

exports[`BiosCard should render extra 1`] = `
<article
  class="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-7"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe="true"
  id=""
>
  <div
    class="pf-c-card__body"
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <div
          class="pf-c-content"
        >
          <h1
            class=""
            data-ouia-component-id="OUIA-Generated-Text-7"
            data-ouia-component-type="PF4/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            BIOS
          </h1>
        </div>
      </div>
      <div
        class="pf-l-stack__item pf-m-fill"
      >
        <div
          class="pf-c-content"
        >
          <dl
            class=""
          >
            <dt
              class=""
            >
              Vendor
            </dt>
            <dd
              class=""
            >
              test-vendor
            </dd>
            <dt
              class=""
            >
              Version
            </dt>
            <dd
              class=""
            >
              test-version
            </dd>
            <dt
              class=""
            >
              Release date
            </dt>
            <dd
              class=""
            >
              01 Apr 2014
            </dd>
            <dt
              class=""
            >
              something
            </dt>
            <dd
              class=""
            >
              test
            </dd>
            <dt
              class=""
            >
              with click
            </dt>
            <dd
              class=""
            >
              <a
                href="http://localhost:5000//undefined"
              >
                1 tests
              </a>
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</article>
`;
